tests/cpp/common/test_host_device_vector.cu

```diff
@@ -256,6 +280,46 @@ TEST(HostDeviceVector, MGPU_Reshard) {
   ASSERT_EQ(total_size, h_vec.size());
   ASSERT_EQ(total_size, vec.Size());
 }
+
+TEST(HostDeviceVector, MGPU_Reshard) {
+  auto devices = GPUSet::AllVisible();
+  if (devices.Size() < 2) {
+    LOG(WARNING) << "Not testing in multi-gpu environment.";
+    return;
+  }
+
+  std::vector<int> h_vec (2345);
+  for (size_t i = 0; i < h_vec.size(); ++i) {
+    h_vec[i] = i;
+  }
+  HostDeviceVector<int> vec (h_vec);
+
+  // Data size for each device.
+  std::vector<size_t> devices_size (devices.Size());
+
+  // From CPU to GPUs.
+  vec.Shard(devices);
+  for (size_t i = 0; i < devices.Size(); ++i) {
+    auto span = vec.DeviceSpan(i);  // sync to device
+  }
+  PlusOne(&vec);
+
+  // Reshard is allowed for already sharded vector.
+  vec.Reshard(GPUDistribution::Overlap(devices, 7));
+  size_t total_size = 0;
+  for (size_t i = 0; i < devices.Size(); ++i) {
+    total_size += vec.DeviceSize(i);
+    devices_size[i] = vec.DeviceSize(i);
+  }
+  size_t overlap = 7 * (devices.Size() - 1);
+  ASSERT_EQ(total_size, h_vec.size() + overlap);
+  ASSERT_EQ(total_size, vec.Size() + overlap);
```

Comment:
Aren't you going to check the data on the device?

Reply:
Changed the test to reuse existing helper methods.